# GAGA Gene family evolution

We used [CAFE](https://github.com/hahnlab/CAFE5) to analyze changes in gene family size across the ant phylogeny. This program uses a birth and death process to model gene gain and loss across the phylogenetic tree. Briefly, we used CAFE to estimate gene family evolutionary rates, both global (the whole phylogeny share the same _birth-and-death_ rate) and local (we tested different scenarios in which we model different _birth-and-death_ rates in the phylogeny, i.e.: one rate for poneroids and other for formicoids). In addition, we infered gene family counts for each orthologous group in all internal nodes of the phylogeny, allowing us to retrieve gene families that have expanded or contracted in specific lineages. 


### CAFE input
First, we use as input in CAFE the orthogroup table that we retrieved in the steps detailed in [Orthology](../Orthology). This table contains the gene count of each infered orthogroup for each species, including our 163 ant genomes and the 8 outgroups from Apoidea. The format of the table should be (example with three genomes):
```
Desc    HOG     GAGA-0001   GAGA-0003   GAGA-0004   
HOG0000754  HOG0000754  5   6   4  
HOG0000866  HOG0000866  3   4   4  
```

In addition, we also use the calibrated phylogeny that contains the phylogenetic relationships and dates retrieved for the 163 ant species sequenced in GAGA, see [Phylogeny](../Phylogeny) for more details on these steps. For the 8 Apoidea species used as outgroups, we reconstructed the phylogeny using these outgroups and single copy ortholog groups, including the 163 ant genomes, and used the dates from [Peters et al. 2017](http://dx.doi.org/10.1016/j.cub.2017.01.027). 

Note that we used the error model implemented in CAFE, which accounts for non-biological factors (e.g., genome sequencing and coverage differences, gene family clustering errors, etc.) leading to incorrect gene family counts in input files. However, we opted to conduct the CAFE analyses both keeping and discarding all species with low genome contiguity (Scaffold N50 < 500 Kb), as the gene family numbers might be biased due to the fragmentation of the assembly, resulting for instance in multiple partial genes or missed copies. And therefore we evaluated both approaches. 
- We removed the columns for the species with low contiguity genomes in the orthogroup gene count table: [N0_GeneCounts_nostLFR.tsv file](N0_GeneCounts_nostLFR.tsv.zip). 
- We pruned the dated species tree to generate a tree without the low contiguity genome species using ***nw_prune*** from [Newick Utilities](https://bio.tools/newick_utilities): [GAGA_dated_phylogeny_newick_woutgroup_nostLFR.tre file](GAGA_dated_phylogeny_newick_woutgroup_nostLFR.tre).
```
nw_prune GAGA_dated_phylogeny_newick_woutgroup.tre GAGA-0002 GAGA-0055 GAGA-0220 GAGA-0234 GAGA-0331 GAGA-0366 GAGA-0386 GAGA-0387 GAGA-0389 GAGA-0392 GAGA-0406 GAGA-0408 GAGA-0495 GAGA-0535 GAGA-0536 GAGA-0537 GAGA-0553 GAGA-0554 GAGA-0577 > GAGA_dated_phylogeny_newick_woutgroup_nostLFR.tre
```

### Running CAFE
We used the script [run_cafe_analyses.sh](run_cafe_analyses.sh) to run the CAFE analyses. The script generates a number of independent runs in order to ensure convergence. It contains the code to run all the following steps that we conducted, just comment/uncomment the respective lines to run each step. 

- I - First, we used all orthogroups to run CAFE with a global lambda model and using a poisson distribution for the root frequency distribution. However, given the number of orthogroups with a high variable number of genes across species, and the large number of species in the phylogeny, the program did not converge. 
- II - Then, we filtered gene families with large gene copy number variance, as these can cause parameter estimates to be non-informative. We tried different values, finally filtering gene families for which the difference in gene number between the genomes with less and most copies were higher than 20. This number can be modified and the filtering conducted using the script [filter_genefams_forcafe.pl](filter_genefams_forcafe.pl). Then, we run CAFE using the ortholog table filtering these orthogroups and ensured that CAFE converged. We checked and evaluated the Maximum-likelihood and estimated lambda(s) of each run using the following command:
```
grep 'L' CAFE*/*/Base_results.txt > Results_lnL_lambda.txt
```
- III - Third, we use the error model in CAFE to estimate a global lambda accounting for non-biological factors. The estimated lambda using the error model is 0.0048 (Vs 0.0050 without the error model)|**Check with the new run**|. This is in accordance with the fact that error in genome assembly and gene annotation should artefactually inflate the rate of gene family evolution, and therefore controlling for it should lead to smaller estimates of lambda. In our case, the difference is really small suggesting the good quality of the genomes and annotation. 
- IV - Fourth, we tested different lambda models (e.g. we compared a three lambda model, having separate rates for the outgroup species, formicoid and poneroid ants) using the error model estimated in the previous step. Note that the models that included more than three lambdas never converged (e.g. one separate rate per each ant subfamily). In this step, we created a tree with the number of lambdas specified within the newick file. The best fitting model was assessed by conducting likelihood ratio tests between the Maximum likelihood scores of the models, and the difference in the parameters is the different number of lambda (function lr.test() in the R library library("extRemes")).

- V - Finally, we used the lambda estimated under the best fitting model, using the error model, to analyze all gene families, including the high variance ones that were filtered to estimate the lambda in the previous steps. 


### Visualizing CAFE results


We plotted the number of expansions and contractions in the phylogeny using the output generated by CAFE with the script cafetutorial_draw_tree.py provided in CAFE tutorial repository: https://github.com/hahnlab/cafe_tutorial
Note that the script requires the phylogenetic tree used in CAFE, as well as the phylogeny with the labels that CAFE uses. These labels can be extracted from the tree in CAFE output named Base_asr.tre. In addition, the header of 'Base_clade_results.txt' output from CAFE need to be renamed as "Node Expansions Contractions".
```
python ../bin/cafetutorial_draw_tree.py -i Base_clade_results.txt -t '(GAGA-0392:156.7416437,(((((((GAGA-0025:46.6011,(GAGA-0352:7.1293,GAGA-0380:7.1293):39.4718):4.75938,((((GAGA-0074:10.6183,NCBI-0011:10.6183):12.8069,GAGA-0353:23.4252):6.85475,GAGA-0301:30.27995):18.8859875,GAGA-0246:49.1659375):2.1945425):9.2664075,((GAGA-0303:3.5554,NCBI-0009:3.5554):48.5748,GAGA-0307:52.1302):8.4966875):1.919770076,((GAGA-0063:36.3075,NCBI-0007:36.3075):22.261725,((GAGA-0083:2.6795,GAGA-0351:2.6795):44.12635,(GAGA-0266:4.1233,GAGA-0300:4.1233):42.68255):11.763375):3.977432576):33.36907478,GAGA-0090:95.91573235):25.4082982,(GAGA-0580:116.8840667,((GAGA-0391:44.4282,GAGA-0404:44.4282):69.43025,GAGA-0552:113.85845):3.025616667):4.439963889):22.03004158,(((GAGA-0379:52.4715,NCBI-0001:52.4715):12.154,GAGA-0534:64.6255):58.48704832,(((((GAGA-0302:14.3932,GAGA-0363:14.3932):16.9916,NCBI-0010:31.3848):27.4284,((GAGA-0337:11.139,GAGA-0340:11.139):31.4595,GAGA-0528:42.5985):16.2147):50.7776,((GAGA-0343:60.1468,GAGA-0365:60.1468):39.9999,(GAGA-0517:4.2804,(GAGA-0521:1.0074,GAGA-0522:1.0074):3.273):95.8663):9.4441):5.196253241,((((((GAGA-0020:1.8828,GAGA-0024:1.8828):10.56007,((GAGA-0177:4.3446,GAGA-0532:4.3446):3.43915,(GAGA-0364:3.0053,(GAGA-0491:0.2215,GAGA-0524:0.2215):2.7838):4.77845):4.65912):0.87853,GAGA-0026:13.3214):42.0995,(((GAGA-0198:22.4344,GAGA-0275:22.4344):3.7093,GAGA-0332:26.1437):14.4307,GAGA-0341:40.5744):14.8465):16.40631087,((((GAGA-0304:27.39435,((GAGA-0334:4.5078,GAGA-0494:4.5078):9.7184,GAGA-0354:14.2262):13.16815):21.68286964,(((GAGA-0350:1.4076,GAGA-0359:1.4076):5.8996,((GAGA-0485:4.9908,NCBI-0008:4.9908):1.4487,(OUT-0001:2.6427,OUT-0002:2.6427):3.7968):0.8677):6.92235,GAGA-0502:14.22955):34.84766964):12.10198945,((((GAGA-0200:16.556,((((GAGA-0221:7.7079,GAGA-0361:7.7079):0.96,GAGA-0362:8.6679):1.6635,NCBI-0005:10.3314):1.5217,GAGA-0396:11.8531):4.7029):13.1445,GAGA-0374:29.7005):10.8006,GAGA-0360:40.5011):16.7532,GAGA-0336:57.2543):3.924909091):0.7155459091,((GAGA-0187:11.5244,GAGA-0338:11.5244):36.2856,GAGA-0199:47.81):14.084755):9.93245587):35.63209764,(GAGA-0306:102.3357986,(((((GAGA-0028:9.315,GAGA-0087:9.315):0.60985,GAGA-0401:9.92485):31.30671667,GAGA-0114:41.23156667):28.65650833,NCBI-0012:69.888075):4.071405,((((((GAGA-0004:4.9121,GAGA-0413:4.9121):8.8861,GAGA-0084:13.7982):0.6099,(GAGA-0356:13.1803,GAGA-0531:13.1803):1.2278):25.61145,(GAGA-0503:15.0019,GAGA-0505:15.0019):25.01765):8.834685714,GAGA-0109:48.85423571):15.81624808,(((((((GAGA-0082:46.6008,(GAGA-0103:41.5199,GAGA-0256:41.5199):5.0809):3.0905,((GAGA-0098:8.47305,(GAGA-0099:6.4548,GAGA-0463:6.4548):2.01825):27.40467083,(GAGA-0222:18.66178571,((((GAGA-0223:7.01285,(GAGA-0512:6.4248,GAGA-0513:6.4248):0.58805):3.951616667,GAGA-0224:10.96446667):0.8777083333,GAGA-0511:11.842175):2.384905,(GAGA-0288:12.2045,GAGA-0510:12.2045):2.02258):4.434705714):17.21593512):13.81357917):1.01325,GAGA-0407:50.70455):0.8572533333,(GAGA-0165:49.6148,NCBI-0013:49.6148):1.947003333):0.2170496078,((GAGA-0330:24.3849,(GAGA-0393:1.1849,GAGA-0395:1.1849):23.2):20.9816,GAGA-0335:45.3665):6.412352941):0.8213214636,((((((GAGA-0328:7.269,GAGA-0382:7.269):9.2324,GAGA-0579:16.5014):13.35305,(GAGA-0378:16.1763,GAGA-0578:16.1763):13.67815):9.25738,GAGA-0533:39.11183):11.996795,GAGA-0520:51.108625):1.008996429,(((GAGA-0333:15.9169,(GAGA-0527:8.9814,GAGA-0085:8.9814):6.9355):1.2652,GAGA-0405:17.1821):28.8186,GAGA-0515:46.0007):6.116921429):0.4825529762):9.318536667,(((GAGA-0245:43.67875,(GAGA-0454:22.2374,NCBI-0002:22.2374):21.44135):5.458216667,GAGA-0346:49.13696667):11.58909015,((((GAGA-0080:11.24945,((GAGA-0376:7.5179,GAGA-0530:7.5179):2.997,GAGA-0384:10.5149):0.73455):0.3732625,GAGA-0358:11.6227125):37.0836975,NCBI-0014:48.70641):5.815738958,(GAGA-0229:53.34899,(GAGA-0540:48.75244286,(GAGA-0543:43.62199231,(GAGA-0541:37.41019167,(NCBI-0006:31.22709091,(GAGA-0539:28.96369,(((GAGA-0538:14.3747,GAGA-0550:14.3747):2.0578,NCBI-0015:16.4325):7.715885714,(NCBI-0017:18.63178333,(NCBI-0016:15.69954,((NCBI-0003:4.1826,NCBI-0004:4.1826):9.147483333,(GAGA-0014:6.87925,(GAGA-0001:1.5644,GAGA-0003:1.5644):5.31485):6.450833333):2.369456667):2.932243333):5.516602381):4.815304286):2.263400909):6.183100758):6.211800641):5.130450549):4.596547143):1.173158958):6.20390786):1.192654254):2.75177272):9.288996208):28.37631861):5.123509902):7.327744728):8.325495079):20.24152381):13.38757162)' -d '(GAGA-0392<287>,(((((((GAGA-0025<225>,(GAGA-0352<197>,GAGA-0380<196>)<224>)<247>,((((GAGA-0074<155>,NCBI-0011<154>)<177>,GAGA-0353<176>)<195>,GAGA-0301<194>)<223>,GAGA-0246<222>)<246>)<261>,((GAGA-0303<221>,NCBI-0009<220>)<245>,GAGA-0307<244>)<260>)<271>,((GAGA-0063<243>,NCBI-0007<242>)<259>,((GAGA-0083<219>,GAGA-0351<218>)<241>,(GAGA-0266<217>,GAGA-0300<216>)<240>)<258>)<270>)<279>,GAGA-0090<278>)<283>,(GAGA-0580<277>,((GAGA-0391<257>,GAGA-0404<256>)<269>,GAGA-0552<268>)<276>)<282>)<285>,(((GAGA-0379<267>,NCBI-0001<266>)<275>,GAGA-0534<274>)<281>,(((((GAGA-0302<215>,GAGA-0363<214>)<239>,NCBI-0010<238>)<255>,((GAGA-0337<213>,GAGA-0340<212>)<237>,GAGA-0528<236>)<254>)<265>,((GAGA-0343<235>,GAGA-0365<234>)<253>,(GAGA-0517<233>,(GAGA-0521<211>,GAGA-0522<210>)<232>)<252>)<264>)<273>,((((((GAGA-0020<175>,GAGA-0024<174>)<193>,((GAGA-0177<153>,GAGA-0532<152>)<173>,(GAGA-0364<151>,(GAGA-0491<129>,GAGA-0524<128>)<150>)<172>)<192>)<209>,GAGA-0026<208>)<231>,(((GAGA-0198<171>,GAGA-0275<170>)<191>,GAGA-0332<190>)<207>,GAGA-0341<206>)<230>)<251>,((((GAGA-0304<169>,((GAGA-0334<127>,GAGA-0494<126>)<149>,GAGA-0354<148>)<168>)<189>,(((GAGA-0350<125>,GAGA-0359<124>)<147>,((GAGA-0485<101>,NCBI-0008<100>)<123>,(OUT-0001<99>,OUT-0002<98>)<122>)<146>)<167>,GAGA-0502<166>)<188>)<205>,((((GAGA-0200<121>,((((GAGA-0221<47>,GAGA-0361<46>)<63>,GAGA-0362<62>)<77>,NCBI-0005<76>)<97>,GAGA-0396<96>)<120>)<145>,GAGA-0374<144>)<165>,GAGA-0360<164>)<187>,GAGA-0336<186>)<204>)<229>,((GAGA-0187<185>,GAGA-0338<184>)<203>,GAGA-0199<202>)<228>)<250>)<263>,(GAGA-0306<249>,(((((GAGA-0028<143>,GAGA-0087<142>)<163>,GAGA-0401<162>)<183>,GAGA-0114<182>)<201>,NCBI-0012<200>)<227>,((((((GAGA-0004<95>,GAGA-0413<94>)<119>,GAGA-0084<118>)<141>,(GAGA-0356<117>,GAGA-0531<116>)<140>)<161>,(GAGA-0503<139>,GAGA-0505<138>)<160>)<181>,GAGA-0109<180>)<199>,(((((((GAGA-0082<61>,(GAGA-0103<45>,GAGA-0256<44>)<60>)<75>,((GAGA-0098<43>,(GAGA-0099<33>,GAGA-0463<32>)<42>)<59>,(GAGA-0222<41>,((((GAGA-0223<15>,(GAGA-0512<9>,GAGA-0513<8>)<14>)<21>,GAGA-0224<20>)<27>,GAGA-0511<26>)<31>,(GAGA-0288<25>,GAGA-0510<24>)<30>)<40>)<58>)<74>)<93>,GAGA-0407<92>)<115>,(GAGA-0165<91>,NCBI-0013<90>)<114>)<137>,((GAGA-0330<89>,(GAGA-0393<73>,GAGA-0395<72>)<88>)<113>,GAGA-0335<112>)<136>)<159>,((((((GAGA-0328<39>,GAGA-0382<38>)<57>,GAGA-0579<56>)<71>,(GAGA-0378<55>,GAGA-0578<54>)<70>)<87>,GAGA-0533<86>)<111>,GAGA-0520<110>)<135>,(((GAGA-0333<69>,(GAGA-0527<53>,GAGA-0085<52>)<68>)<85>,GAGA-0405<84>)<109>,GAGA-0515<108>)<134>)<158>)<179>,(((GAGA-0245<107>,(GAGA-0454<83>,NCBI-0002<82>)<106>)<133>,GAGA-0346<132>)<157>,((((GAGA-0080<67>,((GAGA-0376<37>,GAGA-0530<36>)<51>,GAGA-0384<50>)<66>)<81>,GAGA-0358<80>)<105>,NCBI-0014<104>)<131>,(GAGA-0229<103>,(GAGA-0540<79>,(GAGA-0543<65>,(GAGA-0541<49>,(NCBI-0006<35>,(GAGA-0539<29>,(((GAGA-0538<13>,GAGA-0550<12>)<19>,NCBI-0015<18>)<23>,(NCBI-0017<17>,(NCBI-0016<11>,((NCBI-0003<5>,NCBI-0004<4>)<7>,(GAGA-0014<3>,(GAGA-0001<1>,GAGA-0003<0>)<2>)<6>)<10>)<16>)<22>)<28>)<34>)<48>)<64>)<78>)<102>)<130>)<156>)<178>)<198>)<226>)<248>)<262>)<272>)<280>)<284>)<286>)<288>' -y Expansions -o summary_run1_tree_expans.pdf
````

We also ran [CAFE v4.2.1](https://github.com/hahnlab/CAFE) with the same input, parameters and fixed lambda(s) in order to generate a compatible output with [CAFE_fig](https://github.com/LKremer/CAFE_fig) in order to visualize the CAFE results. 




