# GAGA Analyses

We include here the scripts and methods used in the comparative genomics analyses.

The phylogenetic analyses aimed to reconstruct the species tree across our ant sequenced species and outgroups can be found in [Phylogeny](Phylogeny). 

We conducted the orthology assessment of the protein-coding genes in our 163 ant genomes, detailed in [Orthology](Orthology).

The selection analyses including the scripts and methods are included in [Selection](Selection).

We explored gene family evolution using [CAFE](https://github.com/hahnlab/CAFE5) and the scripts can be found in [Gene_family_evolution](Gene_family_evolution).




### Positive selection

The parameters are that there is a difference of 50% of clades with positive selection, or enrihcment in more than 50% of the species; or enrichment in more than 40% when test or reference have only a low number of counts (i.e.: < number of independent clades /2). also, the candidates need to be under positive selection in at least 40% of the independent clades. 